hasura/graphql-engine
# add gzip brotli compression to http responses (close #2674)

### Description



The server compresses the responses (success only) from `/v1/query` and `/v1/graphql` endpoints. The server chooses either [brotli](https://en.wikipedia.org/wiki/Brotli) or [gzip](https://en.wikipedia.org/wiki/Gzip) based on `Accept-Encoding` client header. If both are requested then the server picks `brotli`. 

### Affected components 


- Server
- Docs
- Build System (tests)
- Tests

### Related Issues



close #2674 

### Solution and Design



- Add a new `Hasura.Server.Compression` module which exposes the compression type and function to compress. 
- Compress the end response (if required) and log the details of compression.

### New things
Compression details are added to the logs. In the start-up log, the new field `enable_compression` is introduced which is a boolean value.

A new `content_encoding` field is added to `http_info` of `http-log`. It is either `brotli` or `gzip` and `null` if the encoding is not enabled or not requested by the client. See the example below.

```json
{
  "timestamp": "2019-08-23T02:05:28.059-05:00",
  "level": "info",
  "type": "http-log",
  "detail": {
    "operation": {
      "query_execution_time": 0.018795933,
      "user_vars": {
        "x-hasura-role": "admin"
      },
      "error": null,
      "request_id": "6283add2-db7b-4eec-9b1e-d14781893b2f",
      "response_size": 45,
      "query": null
    },
    "http_info": {
      "status": 200,
      "http_version": "HTTP/1.1",
      "url": "/v1/query",
      "ip": "127.0.0.1",
      "method": "POST",
      "content_encoding": "brotli"
    }
  }
}
```

### Steps to test and verify


In the console, make any query with the `network` tab opened in developer options. Check the size of response received when compression enabled and disabled.

### Limitations, known bugs & workarounds
